Add rotation controls to editor selection box corners

frenzibyte opened the pull request:
Closes https://github.com/ppy/osu/issues/12433

Replaces the current simple top-centred rotation control with small rotation controls on the corners of the selection box, same as how other applications have it:

https://user-images.githubusercontent.com/22781491/115982157-900cae80-a5a1-11eb-8d4f-daa38ce92cfc.mp4

Since `SelectionBoxDragHandleButton` was made with having the same appearance and handling logic as the regular drag handler in mind, I've changed the abstraction of the selection box controls to the following:
 - (abstract) `SelectionBoxElement`: contains the base appearance and shared logic of all selection box controls.
   - (sealed) `SelectionBoxButton`: simple implementation of buttons for selection boxes.
   - (abstract) `SelectionBoxDragHandle`: contains simple logic for handling mouse dragging.
     - `SelectionBoxScaleHandle`: a type of drag handle used for scaling selected content.
     - `SelectionBoxRotationHandle`: a type of drag handle used for rotating selected content.
     
Felt kinda overengineered at first but I completely wasn't sure about following the old abstraction, so I've made it more sensible with the new controls.

Few notes:
 - I really didn't want to use the undo/redo icons to represent cornered rotation handles, but I've searched for the usual "two-way curved arrows" on the fontawesome website, and got no results or anything similar (tried 45°-rotated two-way straight arrows but felt weird so I've disregarded it).
 - The current logic of having all handles call `updateRotationHandlesVisibility` was the best I can do, I hoped not to have to query which handle is operated or hovered on every update call, but I've come to a conclusion that it's the simplest way to achieve the intended behaviour.
 - I wasn't really fine with having to expose `HoverGained` and `HoverLost` on the selection box controls, but the alternative way of overriding `ReceivePositionalInputAt` to `handles.Any(h => h.ReceivePositionalInputAt(pos))` doesn't quite hold up when moving mouse from one handle to another immediately (with touchscreen for example), as the composite would still be considered hovered and therefore not call the update rotation visibility state method.

frenzibyte committed:
Abstract base appearence and hover update from drag handles

frenzibyte committed:
Change inheritance of selection box buttons to base control instead

frenzibyte committed:
Add rotation drag handle component

frenzibyte committed:
Add composite managing display of selection box drag handles

frenzibyte committed:
Replace separated top-centered rotation button with rotation drag handles

frenzibyte committed:
Add test coverage

frenzibyte committed:
Use colour fade transform for selection box controls

To become harminous with the fade transforms of the rotation control

frenzibyte committed:
Add property for tracking whether control is during operation

bdach commented:
* Is this related in any way to the framework change (https://github.com/ppy/osu-framework/pull/4401)? Did you forget to mark this as dependent on that PR, or is that just a random unrelated change?
* Newly added tests have failed CI. Seems like you did not attempt to run them headless.

bdach commented on osu.Game/Screens/Edit/Compose/Components/SelectionBoxDragHandleDisplay.cs:
```diff
@@ -0,0 +1,103 @@
+
+        private void bindDragHandle(SelectionBoxDragHandle handle)
+        {
+            handle.HoverGained += updateRotationHandlesVisibility;
+            handle.HoverLost += updateRotationHandlesVisibility;
+            handle.OperationStarted += updateRotationHandlesVisibility;
+            handle.OperationEnded += updateRotationHandlesVisibility;
```
I'm confused. So it's possible to have `AddScaleHandle()` call `bindDragHandle()` which in turn calls `updateRotationHandlesVisibility()` on events? The naming is all over the place.

frenzibyte replied:
yeah I was a bit afraid of the naming, both drag handles should call this method to update the visibility of the rotation handles: 
 - when a drag handle is hovered, show the corresponding rotation control
 - when a rotation handle is hovered, show itself.

bdach commented on osu.Game/Screens/Edit/Compose/Components/SelectionBoxDragHandleDisplay.cs:
```diff
@@ -0,0 +1,103 @@
+    /// <summary>
+    /// Represents a display composite containing and managing the visibility state of the selection box's drag handles.
+    /// </summary>
+    public class SelectionBoxDragHandleDisplay : CompositeDrawable
```
not sold on this name. maybe `SelectionBoxDragHandleContainer` to clarify what's the difference between this and a `SelectionBoxDragHandle`

frenzibyte replied:
Wasn't sure about using the suffix `Container` on a composite at first, but I guess it's fine.

bdach commented on osu.Game/Screens/Edit/Compose/Components/SelectionBox.cs:
```diff
@@ -186,75 +186,76 @@ private void recreate()
+            var handle = new SelectionBoxRotationHandle
+            {
+                Anchor = anchor,
+                HandleDrag = e => OnRotation?.Invoke(convertDragEventToAngleOfRotation(e))
```
I'm a bit weirded out by how this works UX-wise now. Feels unintuitive and worse than before. It's probably because the selection box does not rotate itself and always stays upright, and the drag handle responsible for rotation always stays anchored in the same corner it used to be in. For those two reasons it's really difficult to predict what a drag will do.

frenzibyte replied:
Yeah not sure how to improve it on this side without basically looking at rotating the selection box already, I guess if this is not really good to go without rotating then this can be blocked or otherwise.

frenzibyte replied:
I'll take a look at making the selection box rotate visually with the content.

frenzibyte replied:
As an update, I've looked into it and it isn't really possible to support it perfectly without basically refactoring the entire selection handler scaling and rotation logic, so I'm leaving it as-is for now.

bdach reviewed (changes requested):
mostly superficial review since this needs work anyways

frenzibyte commented on osu.Game/Screens/Edit/Compose/Components/SelectionBoxControl.cs:
```diff
@@ -0,0 +1,115 @@
+        private Circle circle;
+
+        /// <summary>
+        /// Whether an operation has began from this control.
```
yep.

frenzibyte committed:
Improve English

Co-authored-by: Bartłomiej Dach <dach.bartlomiej@gmail.com>

frenzibyte commented on osu.Game/Screens/Edit/Compose/Components/SelectionBoxDragHandleDisplay.cs:
```diff
@@ -0,0 +1,103 @@
+            if (activeHandle?.InOperation == true || activeHandle?.IsHovered == true)
+                return;
+
+            displayedRotationHandle?.FadeOut(SelectionBoxControl.TRANSFORM_DURATION, Easing.OutQuint);
+            displayedRotationHandle = null;
```
Thought about that at first but got completely unsure since `AlwaysPresent` is set to true, I guess it's fine.

bdach replied:
i'm not sure what the issue is there...?

frenzibyte replied:
yeah actually nevermind, will apply the change.

frenzibyte replied:
actually seems `VisibilityContainer` have logic to not propagate input when hidden, would you rather overriding that or just fade as I'm currently doing above.

![CleanShot 2021-04-25 at 20 06 39](https://user-images.githubusercontent.com/22781491/116002295-c712ac80-a601-11eb-90e5-b74c62d31e83.png)

bdach replied:
fine to leave as is i suppose

frenzibyte commented on osu.Game/Screens/Edit/Compose/Components/SelectionBoxControl.cs:
```diff
@@ -0,0 +1,115 @@
+        public event Action OperationStarted;
+        public event Action OperationEnded;
+
+        internal event Action HoverGained;
+        internal event Action HoverLost;
```
hm? `HoverGained` and `HoverLost` are both required to be exposed as such for the display class to bind to them, not sure what alternative way to achieve that?

frenzibyte commented on osu.Game/Screens/Edit/Compose/Components/SelectionBoxControl.cs:
```diff
@@ -0,0 +1,115 @@
+        /// <summary>
+        /// Whether this control is currently handling mouse down input.
+        /// </summary>
+        protected bool HandlingMouse { get; set; }
```
`SelectionBoxDragControl` needs to change the value of this property on drag events, if you have a way of doing that without having to expose this then I would do it.

bdach replied:
is that set in `DragEnd` actually needed? won't `MouseUp` handle it?

frenzibyte replied:
hmm, right doesn't seem to be required, will change as such then.

frenzibyte commented:
> * Is this related in any way to the framework change ([ppy/osu-framework#4401](https://github.com/ppy/osu-framework/pull/4401))? Did you forget to mark this as dependent on that PR, or is that just a random unrelated change?

Just a change that I thought I may need, but not really anymore.

frenzibyte committed:
Instantiate selection box on SetUp

frenzibyte committed:
Inherit `VisibilityContainer` and make duration constant protected

frenzibyte committed:
SelectionBoxDragHandleDisplay -> SelectionBoxDragHandleContainer

frenzibyte committed:
Remove leftover step

frenzibyte committed:
Remove protected expose of `HandlingMouse` setter

Regardless of `OnDragEnd()`, `OnMouseUp()` will still be called resetting the value of that state back.

frenzibyte commented:
Blocking until I figure out how to fix the failing test, currently fails due to:
![image](https://user-images.githubusercontent.com/22781491/116002593-258c5a80-a603-11eb-85b9-7bd9832906c1.png)

frenzibyte committed:
Revert "Inherit `VisibilityContainer` and make duration constant protected"

This reverts commit c58ef4230da768aad729e4785a156c80e83bad54. Uhh, how did I push this..

frenzibyte committed:
Add more distance between each hit object in editor selection test

To avoid potentially hovering over the rotation control instead of wherever the test desired to move the mouse to.

peppy committed:
Merge branch 'master' into corner-rotation-controls

peppy commented:
On an initial visual pass:

- The rotation controls should disappear when dragging a scale operation.
- The corners flash when they appear (they shouldn't)
![20210503 180944 (dotnet)](https://user-images.githubusercontent.com/191335/116859323-cb128000-ac3a-11eb-9466-883bbf3e193a.gif)
- There should be a fade when the controls appear (let's say 100ms outquint)

frenzibyte commented:
> * The rotation controls should disappear when dragging a scale operation.

Hmm, you did tell me to keep them shown when rotating, perhaps I was misreading, but sure can do that.

> * The corners flash when they appear (they shouldn't)
>   ![20210503 180944 (dotnet)](https://user-images.githubusercontent.com/191335/116859323-cb128000-ac3a-11eb-9466-883bbf3e193a.gif)

That would be the `FadeColour` not being initial first time, will fix.

> * There should be a fade when the controls appear (let's say 100ms outquint)

That was actually added already but got removed somewhere during the commits, will add back.

peppy commented:
The *should* show during rotation. I'm saying they should not show when scaling.

frenzibyte commented:
I meant when scaling, mistyped.

frenzibyte committed:
Add back mis-removed fade transform

frenzibyte committed:
Finish transforms on controls load complete

frenzibyte committed:
Move selection box control internal events to drag handles

frenzibyte committed:
Consider drag handles active using mouse down instead of when dragged

frenzibyte committed:
Hide the corresponding rotation handle when holding scale handle

bdach commented:
Needs conflict resolution.

frenzibyte committed:
Merge branch 'master' into corner-rotation-controls

frenzibyte commented:
Should be addressed.

frenzibyte committed:
Fix test scene broken on master

frenzibyte committed:
OnOperation -> TriggerOperation

frenzibyte commented on osu.Game/Screens/Edit/Compose/Components/SelectionBoxControl.cs:
```diff
@@ -0,0 +1,97 @@
+
+        protected void OnOperationStarted() => OperationStarted?.Invoke();
+
+        protected void OnOperationEnded() => OperationEnded?.Invoke();
```
TIL, renamed.

peppy commented on osu.Game/Screens/Edit/Compose/Components/SelectionBoxDragHandleContainer.cs:
```diff
@@ -0,0 +1,109 @@
+    /// <summary>
+    /// Represents a display composite containing and managing the visibility state of the selection box's drag handles.
+    /// </summary>
+    public class SelectionBoxDragHandleContainer : CompositeDrawable
```
Reading all the intertwined logic you have setup in this (and related) classes makes me sleepy.

Also, it's halfway between being flexible and not. For instance, what if we decide the bottom buttons should only display when the bottom centre drag handle is being hovered?

I don't have a solution in mind; just commenting that this is a pretty convoluted setup.

peppy commented:
Tentatively getting this in. I have some pending work on this area of code and will likely make some follow-up changes at that point in time.
